sys/crypto: add ccms

fjmolinas opened the pull request:
### Contribution description

This PR extends CCM to add CCM* (CCMS) . It differs from CCM in that it can be used for messages that require only encryption.

The test cases have been taken from [1] as well.

[1] https://mentor.ieee.org/802.15/dcn/04/15-04-0537-02-004b-formal-specification-ccm-star-mode-operation.pdf

### Testing procedure

- run test
`make -C tests/sys_crypto/ all test`

- take a look at [1] to check the implementation (or better said adaptation of the already existing CCM)

fjmolinas commented:
Build failed only on insufficient memory and doxygen mistake:

```
running './dist/tools/headerguards/check.sh' x
Command output:

	--- sys/include/crypto/modes/ccm_common.h
	+++ sys/include/crypto/modes/ccm_common.h
	@@ -46,5 +46,5 @@
	 }
	 #endif
	 
	-#endif /* CRYPTO_MODES_CCM_H */
	+#endif /* CRYPTO_MODES_CCM_COMMON_H */
	 /** @} */
```

I'll add the boards after review:

```
atmega1284p
derfmega128
mega-xplained
microduino-corerf
msb-430
msb-430h
```

mtausig commented on sys/crypto/modes/ccm.c:
```diff
@@ -11,7 +11,7 @@
  *
  * @file
- * @brief       Crypto mode - counter with CBC-MAC
+ * @brief       Crypto mode - counter with CBC-MAC (CMS) and CMS*
```
Wrong names

mtausig commented on sys/crypto/modes/ccm.c:
```diff
@@ -168,7 +174,17 @@ static inline int _fits_in_nbytes(size_t value, uint8_t num_bytes)
-int cipher_encrypt_ccm(cipher_t *cipher,
+static int _validate_mac_length_ccm(uint8_t mac_length)
+{
+    return (mac_length % 2 != 0 || mac_length < 4 || mac_length > 16);
```
Please add a comment what the valid lengths are (also for the star versions).

mtausig commented on tests/sys_crypto/tests-crypto-modes-ccms.c:
```diff
@@ -0,0 +1,318 @@
+    /* MAC */
+    0x4F, 0xDE, 0x52, 0x90, 0x61, 0xF9, 0xC6, 0xF1
+};
+static const size_t TEST_IEEE_0537R2_3_EXPECTED_LEN = 38;
```
Could you maybe add some more test vectors with longer data. Since the CCM implementation had problems in the past in that area, it would be helpful.

fjmolinas replied:
Does it make sense to add those? They both share the same base code, and those extended length cases are already tested in `tests-crypto-modes-ccms.c`, I could still generate some if you feel its needed.

PeterKietzmann replied:
Why not just add one long sequence... Doesn't the above referenced document provide one?

fjmolinas replied:
Nope, I added all the vectors provided in the document. I added in the last fixup some longer test vectors generated with PyCryptodome.

PeterKietzmann commented on sys/crypto/modes/ccm.c:
```diff
@@ -76,7 +82,9 @@ static int ccm_create_mac_iv(cipher_t *cipher, uint8_t auth_data_len, uint8_t M,
         Reserved   Adata    M_    L_    */
-    M_ = (M - 2) / 2;
+    /* valid M values include 0 for CCMS */
+    M_ = M == 0 ? 0 : (M - 2) / 2;
```
I see that `ccm_create_mac_iv` is just a static function. But anyway, before this change `M_` had a different result `if M=0`

PeterKietzmann replied:
According to [Section 2.3 Point 4.](https://mentor.ieee.org/802.15/dcn/04/15-04-0537-02-004b-formal-specification-ccm-star-mode-operation.pdf):

"The value M=0 corresponds to disabling authenticity, since then the authentication field is the empty string"

-> does it need further action?

fjmolinas replied:
No, I think I didnt't like the `M-2` when M==0, but I can leave it as it was before, or I can return right away if `M=0`.

fjmolinas replied:
> I see that `ccm_create_mac_iv` is just a static function. But anyway, before this change `M_` had a different result `if M=0`

Before it would have been invalid for `M` to be `0`, its not a valid input for `ccm`, and it is checked through `_valid_mac_length_ccm`.

fjmolinas replied:
Skipped it entirely in 1f2698c if `mac_length=0`.

PeterKietzmann commented on sys/crypto/modes/ccm.c:
```diff
@@ -167,8 +175,19 @@ static inline int _fits_in_nbytes(size_t value, uint8_t num_bytes)
+/* Valid mac_length are 4, 6, 8 ... 16 octets */
+static int _validate_mac_length_ccm(uint8_t mac_length)
+{
+    return (mac_length % 2 != 0 || mac_length < 4 || mac_length > 16);
```
Returns `0` if valid and `1` if invalid, right? Feels a bit counter intuitive, maybe add a comment?

fjmolinas replied:
Fixed by returning the intuitive value.

PeterKietzmann commented on sys/crypto/modes/ccm.c:
```diff
@@ -305,3 +316,71 @@ int cipher_decrypt_ccm(cipher_t *cipher,
+                        const uint8_t *input, size_t input_len,
+                        uint8_t *plain)
+{
+    if(_validate_mac_length_ccms(mac_length)){
```
Just to double check, this is the only difference to `cipher_decrypt_ccm` above?

fjmolinas replied:
Yes and no. CCM* specifies as well that:

> The nonce N shall encode the potential values for M such that one can uniquely determine from N the actually used value of M. The exact format of the nonce N is outside the scope of this specification and shall be determined and fixed by the actual implementation environment of the CCM* mode.

This part I did not address, this PR intended only to allow messages requiring only encryption. How to generate that nonce to encode M, and how to decode M from the nonce, was not required for me currently. I would address it latter if needed. But I realize this is not clear in the documentation I will add a comment.

PeterKietzmann replied:
Did you check how this is implemented elsewhere? OpenWSN?

PeterKietzmann commented on tests/sys_crypto/tests-crypto-modes-ccm.c:
```diff
@@ -1113,7 +1138,7 @@ static void test_decrypt_op(const uint8_t *key, uint8_t key_len,
                              nonce, nonce_len, encrypted, encrypted_len, data);
-    TEST_ASSERT_MESSAGE(len > 0, "Decryption failed");
+    TEST_ASSERT_MESSAGE(len >= 0, "Decryption failed");
 
```
Can you double check if this is needed? And if so, make sure that this case is documented for `cipher_decrypt_ccm`?

fjmolinas replied:
This is because if `INPUT_LEN = 0` then the decrypted message would be of length 0, which is not incorrect. In the function documentation it says that on error it returns a negative error code. I can add a message to state that the length can be 0, and make explicit that the `input_length` range start at `0`, is that what you meant?

fjmolinas replied:
Did just that.

PeterKietzmann commented on tests/sys_crypto/tests-crypto-modes-ccms.c:
```diff
@@ -0,0 +1,318 @@
+}
+
+#define do_test_decrypt_op(name) do { \
+        test_decrypt_op(TEST_ ## name ## _KEY, TEST_ ## name ## _KEY_LEN, \
```
I'm not familiar with the new *tests/sys_crypto* structure. Is it possible to share this between *tests-crypto-modes-ccm.c* and *tests-crypto-modes-ccms.c* ?

fjmolinas replied:
The macros could be in a shared header file yes, and I could move the shared `test_decrypt_op()/test_encrypt_op` into a helper source file as well.

PeterKietzmann replied:
Ah, this hasn't been addressed yet.

PeterKietzmann commented on tests/sys_crypto/tests-crypto-modes-ccms.c:
```diff
@@ -0,0 +1,318 @@
+}
+
+
+/* Test length checking in ccm functions. */
```
*ccms

PeterKietzmann commented:
> TBH, I don't quite see the point of this cipher mode.

@mtausig the OpenWSN re-integration in #13824 requires it. This mode is involved in [link layer security in TSCH networks](https://hal.inria.fr/hal-01342664/document). Do you have objections?

mtausig commented:
> > TBH, I don't quite see the point of this cipher mode.
> 
> @mtausig the OpenWSN re-integration in #13824 requires it. This mode is involved in [link layer security in TSCH networks](https://hal.inria.fr/hal-01342664/document). Do you have objections?

No, if there is a valid reason for it I'm OK with it. Maybe add a disclaimer in the documentation to prevent it from beeing used for generic encryption.

fjmolinas commented:
> No, if there is a valid reason for it I'm OK with it. Maybe add a disclaimer in the documentation to prevent it from beeing used for generic encryption.

Added a note in the documentation

>> Just to double check, this is the only difference to cipher_decrypt_ccm above?
> Yes and no. CCM* specifies as well that:
> 
> > The nonce N shall encode the potential values for M such that one can uniquely determine from N the actually used value of M. The exact format of the nonce N is outside the scope of this specification and shall be determined and fixed by the actual implementation environment of the CCM* mode.
> 
> This part I did not address, this PR intended only to allow messages requiring only encryption. How to generate that nonce to encode M, and how to decode M from the nonce, was not required for me currently. I would address it latter if needed. But I realize this is not clear in the documentation I will add a comment.

Add more context on this in the documentation. Note that If at some point the nonce is encoded, this would depend on the application. Since this was out of scope of the specification they just use the `source address`, `frame counter` and `security level`, `OpenWSN` does something similar using `source address` and the `ASN` (absolute slot number).

fjmolinas commented:
> Add more context on this in the documentation. Note that If at some point the nonce is encoded, this would depend on the application. Since this was out of scope of the specification they just use the source address, frame counter and security level, OpenWSN does something similar using source address and the ASN (absolute slot number).

Contiki's implementation also just receives the nonce. But let me take a second look elsewhere.

fjmolinas commented:
> The CCM* mode extends the definition of the original CCM mode, such as to provide for confidentiality-only services, in addition to the other security service options already offered. Moreover, the CCM* mode adapts the original CCM mode such that the resulting mode can be used securely with variable-length authentication tags, rather than fixed-length authentication tags only. Thus, the CCM* mode takes away the disadvantages of the CCM mode pointed out above. For the detailed specification of the CCM* mode of operation, see Section 2.

I looked again into `ccm`, `ccms` documentation and implementations. I now think I agree with @mtausig, there is no interest in adding CCMS as this PR is constructed. Using CCMS for encryption only is not recommended security wise, so its not to be expected that upper layers would have a use for it. Only the nonce part could be of interest, but its not part of the PR. OpenWSN on its side although it uses an implementation of `ccms` similar to this PR (so it allows encryption only), it never uses encryption only mode, for the same security issues.

Also in the case of OpenWSN `oscore` or `iee802154e` specifies how the nonce should be computed, so there is no use case for the second feature that a CCM extension would provide. Since my main use case was the OpenWSN PR, and since it is no longer needed, I'll would like to close this PR, and split the change that adds the test and handling of `input_len=0` 3b4aca6e8e31c945b5079baea94a4556428f2568.

If I come around implementing a scheme for nonce encoding, I would re-open.

fjmolinas commented:
Closed.
